File: 02359. Find Closest Node to Given Two Nodes/solution.js
/**
 * Problem: 2359. Find Closest Node to Given Two Nodes
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 16 ms (Beats 100%)
 */

/**
 * Finds node with minimal maximum distance from both node1 and node2
 *
 * @param {number[]} edges - Directed graph representation
 * @param {number} node1 - First starting node
 * @param {number} node2 - Second starting node
 *
 * @returns {number} - Node index or -1 if none exists
 */
const closestMeetingNode = (edges, node1, node2) => {
  // Get the number of nodes in the graph
  const nodeCount = edges.length

  // Initialize distance arrays for each starting node with -1 (unvisited)
  const distancesFromNode1 = Array(nodeCount).fill(-1),
    distancesFromNode2 = Array(nodeCount).fill(-1)

  // DFS function to calculate distances from a starting node
  const calculateDistances = (startNode, distances) => {
    let currentNode = startNode
    let distance = 0

    // Continue until we reach a leaf node or a previously visited node
    while (currentNode !== -1 && distances[currentNode] === -1) {
      distances[currentNode] = distance++
      currentNode = edges[currentNode]
    }
  }

  // Calculate distances from both starting nodes
  calculateDistances(node1, distancesFromNode1)
  calculateDistances(node2, distancesFromNode2)

  // Track the result node and its minimum maximum distance
  let resultNode = -1
  let minMaxDistance = Infinity

  // Find the node with minimum maximum distance from both starting nodes
  for (let i = 0; i < nodeCount; i++) {
    // Check if node is reachable from both starting nodes
    if (distancesFromNode1[i] !== -1 && distancesFromNode2[i] !== -1) {
      // Get the maximum of the two distances
      const maxDistance = Math.max(distancesFromNode1[i], distancesFromNode2[i])

      // Update result if we found a better candidate
      if (
        maxDistance < minMaxDistance ||
        (maxDistance === minMaxDistance && i < resultNode)
      ) {
        minMaxDistance = maxDistance
        resultNode = i
      }
    }
  }

  return resultNode
}

File: 02359. Find Closest Node to Given Two Nodes/solution.go
/**
 * Problem: 2359. Find Closest Node to Given Two Nodes
 *
 * Difficulty: Medium
 *
 * Language: Go
 *
 * Performance: Runtime - 9 ms (Beats 100%)
 */

func closestMeetingNode(edges []int, node1 int, node2 int) int {
	// Create distance arrays for both starting nodes
	dist1, dist2 := make([]int, len(edges)), make([]int, len(edges))

	// Calculate distances from node1 to all reachable nodes
	distance := 1
	for node1 != -1 {
		if dist1[node1] > 0 {
			break // Stop if we detect a cycle
		}

		dist1[node1] = distance
		distance++
		node1 = edges[node1]
	}

	// Calculate distances from node2 to all reachable nodes
	distance = 1
	for node2 != -1 {
		if dist2[node2] > 0 {
			break // Stop if we detect a cycle
		}

		dist2[node2] = distance
		distance++
		node2 = edges[node2]
	}

	// Find node with minimum maximum distance from both starting nodes
	minDistance := math.MaxInt
	result := -1

	for i := range dist1 {
		if dist1[i] > 0 && dist2[i] > 0 {
			maxDistance := max(dist1[i], dist2[i])

			if maxDistance < minDistance {
				minDistance = maxDistance
				result = i
			}
		}
	}
	
	return result
}